### ISSUE TITLE
Predictions with non intersecting dfms

### ISSUE THREAD

**Issue by ibecav** [kbenoit/quanteda#1136]
Hi,

This was addressed partially in Issue #predict.dfm #46  but I didn't want to reopen it.  I stumbled upon it when I had the same problem which is trying to run a prediction (in my case Naive Bayes using textmodel_nb) where the test dfm has fewer features than the train dfm.  The solution given worked at the time but Quanteda has moved along (thank you @kbenoit) and the exact commands are slightly different now.  I'm passing this along under the category of **documentation** for the next person who has this issue.

Seems there are three approaches possible.

1.  Ken's original suggestion "_I would suggest you create just one dfm consisting of both training and test documents, and then slice out the cases you want for training the model, and apply the same slice in negative to apply the prediction to the test set._"  That methodology is amply documented in several existing options so I won't address it here.

If you have two existing separate dfms which I'll simply call train and test and they have non identical feature sets and you can't go with the the first choice above you can:

2.  Choose to reduce both train and test down to the intersecting feature set using:

```
# what do we have?
train # a dfm with 19,579 documents, 25,616 features (99.9% sparse)
test # a dfm with 8,392 documents, 17,882 features (99.9% sparse)
# reconcile
theintersection <- intersect(featnames(train_dtm), featnames(test_dtm))
head(theintersection) # now contains a list of the common words in training and test
length(theintersection) # 14,484
# take the theintersection list and trim the two dtms
train_dtm <- dfm_select(train_dtm, theintersection, selection = "keep", valuetype = "fixed")
test_dtm <- dfm_select(test_dtm, theintersection, selection = "keep", valuetype = "fixed")
# check our work
setequal(featnames(train_dtm), featnames(test_dtm))
```

or you can:

3.  Use dfm_select to create a newtest out of the original test which is padded with zeros for all the features that are in train but not in the original test.  This will allow the predict command to run successfully. This is different from the explanation given in Issue predict.dfm #46 which I believe has been completely deprecated.

```
newtest_dtm <- dfm_select(test_dtm, train_dtm, valuetype = "fixed", verbose = TRUE)
# kept 14,484 features, padded 7,734 features
setequal(featnames(newtest_dtm), featnames(train_dtm))
```
 
```
# build model the same for both #2 or #3 above
mymodel <- textmodel_nb(train_dtm, y = docvars(train_dtm, "author"), smooth = 1, prior = "docfreq")
mymodel

# make predictions for #2
mypredictions <- predict(mymodel, newdata = test)

# make predictions for #3
mypredictions <- predict(mymodel, newdata = newtest)
```

Hope this saves someone some time in the future.  Thanks again @kbenoit for all you do.

Chuck
@ibecav

**Comment by kbenoit** [comment]
@ibecav this is really useful, thanks. We will save it for use in a vignette on machine learning later.

**Comment by ibecav** [comment]
You're very welcome Ken, least I can do given all the work you and others have put into the package.   --Chuck

**Comment by Monduiz** [comment]
Thank you @ibecav! This helped me out. 

I am wondering about this and if a different workflow with dfm would avoid this problem.
Something like binding rows of train and test sets, doing the preprocessing and dfm and then splitting the rows after doing any normalization on the matrix. I should try to see if this problem comes up then too.

**Comment by kbenoit** [comment]
I think you can easily do that now, by creating a docvar for training that has `NA` for documents you want to exclude from the training set. This works for all of the supervised `textmodel_*()` methods. See for instance the `quanteda.corpora::data_corpus_amicus()` for how this is structured. So, you can create one dfm, and anything with NA in the `y` part of the `textmodel_` argument is not used for the training procedure.

**Comment by Monduiz** [comment]
@kbenoit I will go look at what you indicated. This is something I want to learn into a new workflow.
Preprocessing two times for both test and train is not efficient and I have wanted to take the time to code a better workflow but got used to just copy old code.

**Comment by kbenoit** [comment]
See for instance https://github.com/kbenoit/ITAUR/blob/master/6_advanced/advanced.md#further-analysis-examples. (The **quanteda.data** package has been replaced by **quanteda.corpora**.)
